server.go

```diff
@@ -3096,36 +3096,59 @@ func (s *server) rebroadcastHandler(ctx context.Context) {
 func (s *server) querySeeders(ctx context.Context) {
 	// Add peers discovered through DNS to the address manager.
 	seeders := s.chainParams.Seeders()
-	for _, seeder := range seeders {
-		go func(seeder string) {
-			ctx, cancel := context.WithTimeout(ctx, time.Minute)
-			defer cancel()
+	errs := make(chan error, len(seeders))
+	seed := func(seeder string) {
+		ctx, cancel := context.WithTimeout(ctx, time.Minute)
+		defer cancel()
 
-			addrs, err := connmgr.SeedAddrs(ctx, seeder, dcrdDial,
-				connmgr.SeedFilterServices(defaultRequiredServices))
-			if err != nil {
-				srvrLog.Infof("seeder '%s' error: %v", seeder, err)
-				return
-			}
+		addrs, err := connmgr.SeedAddrs(ctx, seeder, dcrdDial,
+			connmgr.SeedFilterServices(defaultRequiredServices))
+		if err != nil {
+			srvrLog.Infof("seeder '%s' error: %v", seeder, err)
+			errs <- err
+			return
+		}
 
-			// Nothing to do if the seeder didn't return any addresses.
-			if len(addrs) == 0 {
-				return
-			}
+		// Nothing to do if the seeder didn't return any addresses.
+		if len(addrs) == 0 {
+			errs <- nil
+			return
+		}
 
-			// Lookup the IP of the https seeder to use as the source of the
-			// seeded addresses.  In the incredibly rare event that the lookup
-			// fails after it just succeeded, fall back to using the first
-			// returned address as the source.
-			srcAddr := wireToAddrmgrNetAddress(addrs[0])
-			srcIPs, err := dcrdLookup(seeder)
-			if err == nil && len(srcIPs) > 0 {
-				const httpsPort = 443
-				srcAddr = addrmgr.NewNetAddressIPPort(srcIPs[0], httpsPort, 0)
+		// Lookup the IP of the https seeder to use as the source of the
+		// seeded addresses.  In the incredibly rare event that the lookup
+		// fails after it just succeeded, fall back to using the first
+		// returned address as the source.
+		srcAddr := wireToAddrmgrNetAddress(addrs[0])
+		srcIPs, err := dcrdLookup(seeder)
+		if err == nil && len(srcIPs) > 0 {
+			const httpsPort = 443
+			srcAddr = addrmgr.NewNetAddressIPPort(srcIPs[0], httpsPort, 0)
+		}
+		addresses := wireToAddrmgrNetAddresses(addrs)
+		s.addrManager.AddAddresses(addresses, srcAddr)
+		errs <- nil
+	}
+
+	backoff := time.Second
+	for {
+		for _, seeder := range seeders {
+			go seed(seeder)
+		}
+		var errCount int
+		for range seeders {
+			if err := <-errs; err != nil {
+				errCount++
 			}
-			addresses := wireToAddrmgrNetAddresses(addrs)
-			s.addrManager.AddAddresses(addresses, srcAddr)
-		}(seeder)
+		}
+		if errCount < len(seeders) || !s.addrManager.NeedMoreAddresses() {
```

Comment:
Probably need to check for `ctx.Err() != nil` here too.  Otherwise, if the ctx is canceled, all of the grs will fail, and `errCount` will be equal to `len(seeders)` thereby leaking the `querySeeder` goroutine.

Reply:
i considered this. the previous code also spawned goroutines that may still be left running after the server exits.

Reply:
Hmm.  They are only potentially running for a very short period though and thus aren't leaked.  This new code would have the gr run forever (and constantly launch new ones that would fail).

`querySeeders` currently blocks until a goroutine for each seeder is launched and those goroutines all create a child context from the provided context along with a timeout and then use that child context in the call to `connmgr.SeedAddrs` which will error on context cancel (whether that comes from propagation via parent or via timeout).  So, in that case, the error path will be hit and log the error and exit the gr.

In the case it happened to already be done with the call to `connmgr.SeedAddrs`, the only other call that could take any real time would be `dcrdLookup`.  That will still exit though.

As a slight aside, ideally, `dcrdLookup` probably should be taking a context these days too now that we have contexts running throughout the entire lifecycle.  That was all clearly written pre-context.

Reply:
even if they are still running when the program exits, that's not a problem. like i said i looked into if this mattered, and figured it didn't.